Improvements from TheForge

darksylinc opened the pull request:
The work was performed by collaboration of TheForge and Google. I am merely splitting it up into smaller PRs and cleaning it up.

This is the most "risky" PR so far because the previous ones have been miscellaneous stuff aimed at either [improve debugging](https://github.com/godotengine/godot/pull/90993) (e.g. device lost), [improve Android experience](https://github.com/godotengine/godot/pull/96439) (add Swappy for better Frame Pacing + Pre-Transformed Swapchains for slightly better performance), or harmless [ASTC improvements](https://github.com/godotengine/godot/pull/96045) (better performance by simply toggling a feature when available).

However this PR contains larger modifications aimed at improving performance or reducing memory fragmentation. With greater modifications, come greater risks of bugs or breakage.

Changes introduced by this PR:

## Transient memory

TBDR GPUs (e.g. most of Android + iOS + M1 Apple) support rendering to Render Targets that are not backed by actual GPU memory (everything stays in cache). This works as long as load action isn't `LOAD`, and store action must be `DONT_CARE`. This saves VRAM (it also makes painfully obvious when a mistake introduces a performance regression). Of particular usefulness is when doing MSAA and keeping the raw MSAA content is not necessary.

## Immutable samplers

Some GPUs get faster when the sampler settings are hard-coded into the GLSL shaders (instead of being dynamically bound at runtime). This required changes to the GLSL shaders, PSO creation routines, Descriptor creation routines, and Descriptor binding routines.

### Toggle

 - `bool immutable_samplers_enabled = true`

Setting it to false enforces the old behavior. Useful for debugging bugs and regressions.

Immutable samplers requires that the samplers stay... immutable, hence this boolean is useful if the promise gets broken. We might want to turn this into a `GLOBAL_DEF` setting.

## Linear Descriptor Pools

Instead of creating dozen/hundreds/thousands of `VkDescriptorSet` every frame that need to be freed individually when they are no longer needed, they all get freed at once by resetting the whole pool. Once the whole pool is no longer in use by the GPU, it gets reset and its memory recycled. Descriptor sets that are created to be kept around for longer or forever (i.e. not created and freed within the same frame) **must not** use linear pools. There may be more than one pool per frame. How many pools per frame Godot ends up with depends on its capacity, and that is controlled by `rendering/rendering_device/vulkan/max_descriptors_per_pool`.

  - **Possible improvement for later:** It should be possible for Godot to adapt to how many descriptors per pool are needed on a per-key basis (i.e. grow their capacity like `std::vector` does) after rendering a few frames; which would be better than the current solution of having a single global value for all pools (`max_descriptors_per_pool`) that the user needs to tweak.

### Toggle
 - `bool linear_descriptor_pools_enabled = true`

Setting it to false enforces the old behavior. Useful for debugging bugs and regressions. Setting it to false is required when workarounding driver bugs (e.g. Adreno 730).

## Reset Command Pools

A ridiculous optimization. Ridiculous because the original code should've done this in the first place. Previously Godot was doing the following:

  1. Create a command buffer **pool**. One per frame.
  2. Create multiple command buffers from the pool in point 1.
  3. Call `vkBeginCommandBuffer` on the cmd buffer in point 2. This resets the cmd buffer because Godot requests the `VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT` flag.
  4. Add commands to the cmd buffers from point 2.
  5. Submit those commands.
  6. On frame N + 2, recycle the buffer pool and cmd buffers from pt 1 & 2, and repeat from step 3.

The problem here is that step 3 resets each command buffer individually. Initially Godot used to have 1 cmd buffer per pool, thus the impact is very low.

But not anymore (specially with Adreno workarounds to force splitting compute dispatches into a new cmd buffer, more on this later). However Godot keeps around a very low amount of command buffers per frame.

The recommended method is to reset the whole pool, to reset all cmd buffers at once. Hence the new steps would be:

  1. Create a command buffer **pool**. One per frame.
  2. Create multiple command buffers from the pool in point 1.
  3. Call `vkBeginCommandBuffer` on the cmd buffer in point 2, which is already reset/empty (see step 6).
  4. Add commands to the cmd buffers from point 2.
  5. Submit those commands.
  6. On frame N + 2, recycle the buffer pool and cmd buffers from pt 1 & 2, call `vkResetCommandPool` and repeat from step 3.

**Possible issues:** @dariosamo added `transfer_worker` which creates a command buffer pool:

```cpp
transfer_worker->command_pool = driver->command_pool_create(transfer_queue_family, RDD::COMMAND_BUFFER_TYPE_PRIMARY);
```

As expected, validation was complaining that command buffers were being reused without being reset (that's good, we now know Validation Layers will warn us of wrong use). I fixed it by adding:

```cpp
void RenderingDevice::_wait_for_transfer_worker(TransferWorker *p_transfer_worker) {
	driver->fence_wait(p_transfer_worker->command_fence);
	driver->command_pool_reset(p_transfer_worker->command_pool); // ! New line !
```

**Secondary cmd buffers are subject to the same issue but I didn't alter them. I talked this with Dario and he is aware of this.** Secondary cmd buffers are currently disabled due to other issues (it's disabled on master).

### Toggle

 - `bool RenderingDeviceCommons::command_pool_reset_enabled`

Setting it to false enforces the old behavior. Useful for debugging bugs and regressions.

There's no other reason for this boolean. Possibly once it becomes well tested, the boolean could be removed entirely.

## Descriptor set batched binding

Adds `command_bind_render_uniform_sets` and `add_draw_list_bind_uniform_sets` (+ compute variants).

It performs the same as `add_draw_list_bind_uniform_set` (notice singular vs plural), but on multiple consecutive uniform sets, thus reducing graph and draw call overhead.

### Toggle

 - `bool descriptor_set_batching = true;`

Setting it to false enforces the old behavior. Useful for debugging bugs and regressions.

There's no other reason for this boolean. Possibly once it becomes well tested, the boolean could be removed entirely.

## Do not wait so long for swapchain

Godot currently does the following:

 1. Fill the entire cmd buffer with commands.
 2. `submit()`
    - Wait with a semaphore for the swapchain.
    - Trigger a semaphore to indicate when we're done (so the swapchain can submit).
 3. `present()`

The optimization opportunity here is that 95% of Godot's rendering is done offscreen. Then a fullscreen pass copies everything to the swapchain. Godot doesn't practically render directly to the swapchain.

The problem with this is that the GPU has to wait for the swapchain to be released **to start anything**, when we could start *much earlier*. Only the final blit pass must wait for the swapchain.

TheForge changed it to the following (more complicated, I'm simplifying the idea):

 1. Fill the entire cmd buffer with commands.
 2. In `screen_prepare_for_drawing` do `submit()`
    - There are no semaphore waits for the swapchain.
    - Trigger a semaphore to indicate when we're done.
 3. Fill a new cmd buffer that only does the final blit to the swapchain.
 4. `submit()`
    - Wait with a semaphore for the submit() from step 2.
    - Wait with a semaphore for the swapchain (so the swapchain can submit).
    - Trigger a semaphore to indicate when we're done (so the swapchain can submit).
 5. `present()`

Dario discovered this problem independently while working on a different platform.

**However TheForge's solution had to be rewritten from scratch:** The complexity to achieve the solution was high and quite difficult to maintain with the way Godot works now (after Übershaders PR). But on the other hand, re-implementing the solution became much simpler because Dario already had to do something similar: To fix an Adreno 730 driver bug, he had to implement splitting command buffers. **This is exactly what we need!**. Thus it was re-written using this existing functionality for a new purpose.

To achieve this, I added a new argument, `bool p_split_cmd_buffer`, to `RenderingDeviceGraph::add_draw_list_begin`, which is only set to true by `RenderingDevice::draw_list_begin_for_screen`.

The graph will split the draw list into its own command buffer.

### Toggle

 - `bool split_swapchain_into_its_own_cmd_buffer = true;`

Setting it to false enforces the old behavior. This might be necessary for consoles which follow an alternate solution to the same problem. If not, then we should consider removing it.

## Free Shader memory

PR #90993 added `shader_destroy_modules()` but it was not actually in use.

This PR adds several places where `shader_destroy_modules()` is called after initialization to free up memory of SPIR-V structures that are no longer needed.

# What's missing?

The following improvements from TheForge were left out from this PR:

1. Render Pass Optimizations.
    - They were mostly good but I need more time to analyze what they do and that they actually help.
2. Replace Push Constants by UBOs.
    - They were not included as they were a bit controversial since point 3 was left out too (continue reading)
3. UMA buffers. They were the most anticipated feature, but unfortunately full of race conditions. It's the feature that needs the most work. Without this feature, it's unclear if replacing push constants by UBOs are a net win or a net loss.

# CI failures

- [x] Deal with DISABLE_DEPRECATED removing `INITIAL_ACTION_CLEAR_REGION` & co.
- [x] Fix D3D12.
- [x] Fix Metal.

darksylinc committed:
Improvements from TheForge (see description)

The work was performed by collaboration of TheForge and Google. I am
merely splitting it up into smaller PRs and cleaning it up.

This is the most "risky" PR so far because the previous ones have been
miscellaneous stuff aimed at either [improve
debugging](https://github.com/godotengine/godot/pull/90993) (e.g. device
lost), [improve Android
experience](https://github.com/godotengine/godot/pull/96439) (add Swappy
for better Frame Pacing + Pre-Transformed Swapchains for slightly better
performance), or harmless [ASTC
improvements](https://github.com/godotengine/godot/pull/96045) (better
performance by simply toggling a feature when available).

However this PR contains larger modifications aimed at improving
performance or reducing memory fragmentation. With greater
modifications, come greater risks of bugs or breakage.

Changes introduced by this PR:

TBDR GPUs (e.g. most of Android + iOS + M1 Apple) support rendering to
Render Targets that are not backed by actual GPU memory (everything
stays in cache). This works as long as load action isn't `LOAD`, and
store action must be `DONT_CARE`. This saves VRAM (it also makes
painfully obvious when a mistake introduces a performance regression).
Of particular usefulness is when doing MSAA and keeping the raw MSAA
content is not necessary.

Some GPUs get faster when the sampler settings are hard-coded into the
GLSL shaders (instead of being dynamically bound at runtime). This
required changes to the GLSL shaders, PSO creation routines, Descriptor
creation routines, and Descriptor binding routines.

 - `bool immutable_samplers_enabled = true`

Setting it to false enforces the old behavior. Useful for debugging bugs
and regressions.

Immutable samplers requires that the samplers stay... immutable, hence
this boolean is useful if the promise gets broken. We might want to turn
this into a `GLOBAL_DEF` setting.

Instead of creating dozen/hundreds/thousands of `VkDescriptorSet` every
frame that need to be freed individually when they are no longer needed,
they all get freed at once by resetting the whole pool. Once the whole
pool is no longer in use by the GPU, it gets reset and its memory
recycled. Descriptor sets that are created to be kept around for longer
or forever (i.e. not created and freed within the same frame) **must
not** use linear pools. There may be more than one pool per frame. How
many pools per frame Godot ends up with depends on its capacity, and
that is controlled by
`rendering/rendering_device/vulkan/max_descriptors_per_pool`.

- **Possible improvement for later:** It should be possible for Godot
to adapt to how many descriptors per pool are needed on a per-key basis
(i.e. grow their capacity like `std::vector` does) after rendering a few
frames; which would be better than the current solution of having a
single global value for all pools (`max_descriptors_per_pool`) that the
user needs to tweak.

 - `bool linear_descriptor_pools_enabled = true`

Setting it to false enforces the old behavior. Useful for debugging bugs
and regressions.
Setting it to false is required when workarounding driver bugs (e.g.
Adreno 730).

A ridiculous optimization. Ridiculous because the original code
should've done this in the first place. Previously Godot was doing the
following:

  1. Create a command buffer **pool**. One per frame.
  2. Create multiple command buffers from the pool in point 1.
3. Call `vkBeginCommandBuffer` on the cmd buffer in point 2. This
resets the cmd buffer because Godot requests the
`VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT` flag.
  4. Add commands to the cmd buffers from point 2.
  5. Submit those commands.
6. On frame N + 2, recycle the buffer pool and cmd buffers from pt 1 &
2, and repeat from step 3.

The problem here is that step 3 resets each command buffer individually.
Initially Godot used to have 1 cmd buffer per pool, thus the impact is
very low.

But not anymore (specially with Adreno workarounds to force splitting
compute dispatches into a new cmd buffer, more on this later). However
Godot keeps around a very low amount of command buffers per frame.

The recommended method is to reset the whole pool, to reset all cmd
buffers at once. Hence the new steps would be:

  1. Create a command buffer **pool**. One per frame.
  2. Create multiple command buffers from the pool in point 1.
3. Call `vkBeginCommandBuffer` on the cmd buffer in point 2, which is
already reset/empty (see step 6).
  4. Add commands to the cmd buffers from point 2.
  5. Submit those commands.
6. On frame N + 2, recycle the buffer pool and cmd buffers from pt 1 &
2, call `vkResetCommandPool` and repeat from step 3.

**Possible issues:** @dariosamo added `transfer_worker` which creates a
command buffer pool:

```cpp
transfer_worker->command_pool =
driver->command_pool_create(transfer_queue_family,
RDD::COMMAND_BUFFER_TYPE_PRIMARY);
```

As expected, validation was complaining that command buffers were being
reused without being reset (that's good, we now know Validation Layers
will warn us of wrong use).
I fixed it by adding:

```cpp
void RenderingDevice::_wait_for_transfer_worker(TransferWorker
*p_transfer_worker) {
	driver->fence_wait(p_transfer_worker->command_fence);
	driver->command_pool_reset(p_transfer_worker->command_pool); //
! New line !
```

**Secondary cmd buffers are subject to the same issue but I didn't alter
them. I talked this with Dario and he is aware of this.**
Secondary cmd buffers are currently disabled due to other issues (it's
disabled on master).

 - `bool RenderingDeviceCommons::command_pool_reset_enabled`

Setting it to false enforces the old behavior. Useful for debugging bugs
and regressions.

There's no other reason for this boolean. Possibly once it becomes well
tested, the boolean could be removed entirely.

Adds `command_bind_render_uniform_sets` and
`add_draw_list_bind_uniform_sets` (+ compute variants).

It performs the same as `add_draw_list_bind_uniform_set` (notice
singular vs plural), but on multiple consecutive uniform sets, thus
reducing graph and draw call overhead.

 - `bool descriptor_set_batching = true;`

Setting it to false enforces the old behavior. Useful for debugging bugs
and regressions.

There's no other reason for this boolean. Possibly once it becomes well
tested, the boolean could be removed entirely.

Godot currently does the following:

 1. Fill the entire cmd buffer with commands.
 2. `submit()`
    - Wait with a semaphore for the swapchain.
- Trigger a semaphore to indicate when we're done (so the swapchain
can submit).
 3. `present()`

The optimization opportunity here is that 95% of Godot's rendering is
done offscreen.
Then a fullscreen pass copies everything to the swapchain. Godot doesn't
practically render directly to the swapchain.

The problem with this is that the GPU has to wait for the swapchain to
be released **to start anything**, when we could start *much earlier*.
Only the final blit pass must wait for the swapchain.

TheForge changed it to the following (more complicated, I'm simplifying
the idea):

 1. Fill the entire cmd buffer with commands.
 2. In `screen_prepare_for_drawing` do `submit()`
    - There are no semaphore waits for the swapchain.
    - Trigger a semaphore to indicate when we're done.
3. Fill a new cmd buffer that only does the final blit to the
swapchain.
 4. `submit()`
    - Wait with a semaphore for the submit() from step 2.
- Wait with a semaphore for the swapchain (so the swapchain can
submit).
- Trigger a semaphore to indicate when we're done (so the swapchain
can submit).
 5. `present()`

Dario discovered this problem independently while working on a different
platform.

**However TheForge's solution had to be rewritten from scratch:** The
complexity to achieve the solution was high and quite difficult to
maintain with the way Godot works now (after Übershaders PR).
But on the other hand, re-implementing the solution became much simpler
because Dario already had to do something similar: To fix an Adreno 730
driver bug, he had to implement splitting command buffers. **This is
exactly what we need!**. Thus it was re-written using this existing
functionality for a new purpose.

To achieve this, I added a new argument, `bool p_split_cmd_buffer`, to
`RenderingDeviceGraph::add_draw_list_begin`, which is only set to true
by `RenderingDevice::draw_list_begin_for_screen`.

The graph will split the draw list into its own command buffer.

 - `bool split_swapchain_into_its_own_cmd_buffer = true;`

Setting it to false enforces the old behavior. This might be necessary
for consoles which follow an alternate solution to the same problem.
If not, then we should consider removing it.

PR #90993 added `shader_destroy_modules()` but it was not actually in
use.

This PR adds several places where `shader_destroy_modules()` is called
after initialization to free up memory of SPIR-V structures that are no
longer needed.

RandomShaper commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -3475,7 +3514,7 @@ Vector<uint8_t> RenderingDeviceDriverVulkan::shader_compile_binary_from_spirv(Ve
 }
 
-RDD::ShaderID RenderingDeviceDriverVulkan::shader_create_from_bytecode(const Vector<uint8_t> &p_shader_binary, ShaderDescription &r_shader_desc, String &r_name) {
+RDD::ShaderID RenderingDeviceDriverVulkan::shader_create_from_bytecode(const Vector<uint8_t> &p_shader_binary, ShaderDescription &r_shader_desc, String &r_name, const Vector<ImmutableSampler> &r_immutableSamplers) {
```
```suggestion
RDD::ShaderID RenderingDeviceDriverVulkan::shader_create_from_bytecode(const Vector<uint8_t> &p_shader_binary, ShaderDescription &r_shader_desc, String &r_name, const Vector<ImmutableSampler> &r_immutable_samplers) {
```

RandomShaper commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -1728,7 +1737,27 @@ RDD::TextureID RenderingDeviceDriverVulkan::texture_create(const TextureFormat &
-	alloc_create_info.preferredFlags = VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT;
+
+	if (p_format.usage_bits & TEXTURE_USAGE_TRANSIENT_BIT) {
+		uint32_t memoryTypeIndex = 0;
```
```suggestion
		uint32_t memory_type_index = 0;
```

RandomShaper commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4365,37 +4413,67 @@ void RenderingDevice::draw_list_draw(DrawListID p_list, bool p_use_indices, uint
-			UniformSet *uniform_set = uniform_set_owner.get_or_null(dl->state.sets[i].uniform_set);
-			_uniform_set_update_shared(uniform_set);
+		if (!dl->state.sets[i].bound) {
+			// @NicolaTF: batch contiguous descriptor sets in a single call
```
```suggestion
			// Batch contiguous descriptor sets in a single call.
```

RandomShaper commented on servers/rendering/rendering_device.cpp:
```diff
@@ -5906,14 +6058,22 @@ uint64_t RenderingDevice::get_memory_usage(MemoryType p_type) const {
 }
 
-void RenderingDevice::_begin_frame() {
+void RenderingDevice::_begin_frame(bool presented) {
```
```suggestion
void RenderingDevice::_begin_frame(bool p_presented) {
```

RandomShaper commented on servers/rendering/rendering_device.h:
```diff
@@ -877,6 +883,7 @@ class RenderingDevice : public RenderingDeviceCommons {
 	RID shader_create_from_spirv(const Vector<ShaderStageSPIRVData> &p_spirv, const String &p_shader_name = "");
 	RID shader_create_from_bytecode(const Vector<uint8_t> &p_shader_binary, RID p_placeholder = RID());
 	RID shader_create_placeholder();
+	void shader_destroy_modules(RID p_shaderRID);
```
```suggestion
	void shader_destroy_modules(RID p_shader);
```

RandomShaper commented on drivers/vulkan/rendering_device_driver_vulkan.h:
```diff
@@ -444,7 +445,15 @@ class RenderingDeviceDriverVulkan : public RenderingDeviceDriver {
 	virtual String shader_get_binary_cache_key() override final;
 	virtual Vector<uint8_t> shader_compile_binary_from_spirv(VectorView<ShaderStageSPIRVData> p_spirv, const String &p_shader_name) override final;
-	virtual ShaderID shader_create_from_bytecode(const Vector<uint8_t> &p_shader_binary, ShaderDescription &r_shader_desc, String &r_name) override final;
+	/** Creates a Pipeline State Object (PSO) out of the shader and all the input data it needs.
```
It's good to have some documentation, but Godot is not following this convention. I don't know what should we do.

darksylinc replied:
There are cases where Godot follows doxygen documentation, such as:

- gdextension_interface.h
- `split_by_plane` in godot/core/math/face3.h

I followed those same conventions.

I know it's rare and Godot doesn't have code comments (which is very bad) and there are no standards set; but the `r_immutableSamplers` parameter was simply important to clarify the gotchas (that caller promises not to bind a different sampler because that simply won't work).

RandomShaper replied:
Yeah, makes sense. Maybe this eventually becomes the standard to follow. In the meantime, there's no real reson to discuss it. I just stroke my eye.

RandomShaper commented on drivers/vulkan/rendering_device_driver_vulkan.h:
```diff
@@ -482,17 +491,25 @@ class RenderingDeviceDriverVulkan : public RenderingDeviceDriver {
 		VkDescriptorPool vk_descriptor_pool = VK_NULL_HANDLE;
-		DescriptorSetPools::Iterator pool_sets_it;
+		VkDescriptorPool vk_linear_descriptor_pool = VK_NULL_HANDLE;
+		DescriptorSetPools::Iterator pool_sets_it = {};
```
```suggestion
		DescriptorSetPools::Iterator pool_sets_it;
```

RandomShaper commented on thirdparty/vulkan/vk_mem_alloc.h:
Changes to a third-party item involve keeping a patch to the original file so the dependency can be "reconstructed" upon upgrading it. For an example, see `thirdparty/spirv_reflect/patches` and the related info about it in `thirdparty/README.md`.

darksylinc replied:
Ok done!

I added `0003-VMA-add-vmaCalculateLazilyAllocatedBytes.patch`.

There was no need to modify `thirdparty/README.md` because VMA already has custom patches and has a section for it.

The only thing I noticed is that the already-existing patches (`VKEnumStringHelper-use-godot-vulkan.patch` and `VMA-use-godot-vulkan.patch`) do not say in which order have to be applied (maybe any order works).

Which is why I named `0003-VMA-add-vmaCalculateLazilyAllocatedBytes.patch` starting with `0003` to indicate it's the 3rd one that needs to be applied.

This convention is used by git if you ask git to generate the patches for you. You can see the only one following this convention is the collection of patches in `thirdparty/vhacd`.

RandomShaper commented on servers/rendering/rendering_device_graph.h:
```diff
@@ -687,17 +705,19 @@ class RenderingDeviceGraph {
 	void add_draw_list_bind_index_buffer(RDD::BufferID p_buffer, RDD::IndexBufferFormat p_format, uint32_t p_offset);
 	void add_draw_list_bind_pipeline(RDD::PipelineID p_pipeline, BitField<RDD::PipelineStageBits> p_pipeline_stage_bits);
 	void add_draw_list_bind_uniform_set(RDD::ShaderID p_shader, RDD::UniformSetID p_uniform_set, uint32_t set_index);
+	void add_draw_list_bind_uniform_sets(RDD::ShaderID p_shader, VectorView<RDD::UniformSetID> p_uniform_set, uint32_t first_index, uint32_t set_count);
```
```suggestion
	void add_draw_list_bind_uniform_sets(RDD::ShaderID p_shader, VectorView<RDD::UniformSetID> p_uniform_set, uint32_t p_first_index, uint32_t p_set_count);
```

RandomShaper commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4600,6 +4697,21 @@ void RenderingDeviceDriverVulkan::command_bind_render_uniform_set(CommandBufferI
+	}
+
+	const ShaderInfo *shader_info = (const ShaderInfo *)p_shader.id;
+	VkDescriptorSet *sets = (VkDescriptorSet *)alloca(sizeof(VkDescriptorSet) * p_set_count);
```
In the driver code we use an utility macro (not repeating the comment for every occurence):

```suggestion
	VkDescriptorSet *sets = ALLOCA_ARRAY(VkDescriptorSet, p_set_count);
```

RandomShaper replied:
For some reason, we're also using the thread-local vector approach in driver code, as stated here: https://github.com/godotengine/godot/pull/99257/files#r1843284766

So, feel free to pick either. There will have to be a uniformity pass in the future.

darksylinc replied:
Done.

Switched to thread_local.

RandomShaper commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4365,37 +4413,67 @@ void RenderingDevice::draw_list_draw(DrawListID p_list, bool p_use_indices, uint
 		}
 	}
 #endif
+	RDD::UniformSetID *valid_descriptor_ids = (RDD::UniformSetID *)alloca(dl->state.set_count * sizeof(RDD::UniformSetID));
```
In the `RenderingDevice` code we use a thread-local vector for this (not reiterating the comment on every case in the PR):

```suggestion
	thread_local LocalVector<RDD::UniformSetID> valid_descriptor_ids;
	valid_descriptor_ids.resize(dl->state.set_count);
```

RandomShaper commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4884,37 +4978,66 @@ void RenderingDevice::compute_list_dispatch(ComputeListID p_list, uint32_t p_x_g
 	}
 
+	// Bind the remaining batch
+	if (valid_set_count > 0)
+		draw_graph.add_compute_list_bind_uniform_sets(cl->state.pipeline_shader_driver_id, VectorView(valid_descriptor_ids, valid_set_count), first_set_index, valid_set_count);
```
```suggestion
	if (valid_set_count > 0) {
		draw_graph.add_compute_list_bind_uniform_sets(cl->state.pipeline_shader_driver_id, VectorView(valid_descriptor_ids, valid_set_count), first_set_index, valid_set_count);
	}
```

RandomShaper commented on servers/rendering/rendering_device.cpp:
```diff
@@ -5962,22 +6122,24 @@ void RenderingDevice::_end_frame() {
-
-	// Execute command buffers and use semaphores to wait on the execution of the previous one. Normally there's only one command buffer,
-	// but driver workarounds can force situations where there'll be more.
+void RenderingDevice::execute_chained_cmds(const bool present_swap_chain, const RDD::FenceID fence,
```
```suggestion
void RenderingDevice::execute_chained_cmds(bool p_present_swap_chain, RDD::FenceID p_fence,
```

Feel free to read the `const` (I guess that ship already sailed), though.

tetrapod00 commented on doc/classes/ProjectSettings.xml:
```diff
@@ -2861,6 +2861,9 @@
 			[b]Note:[/b] Some platforms may restrict the actual value.
 		</member>
 		<member name="rendering/rendering_device/vulkan/max_descriptors_per_pool" type="int" setter="" getter="" default="64">
+			The number of descriptors per pool. Godot's Vulkan backend uses linear pools for descriptors that will be created and destroyed within a single frame. Instead of destroying every single descriptor every frame, they all can be destroyed at once by resetting the pool they belong to.
```
```suggestion
			The number of descriptors per pool. The Vulkan rendering driver uses linear pools for descriptors that will be created and destroyed within a single frame. Instead of destroying every single descriptor every frame, they all can be destroyed at once by resetting the pool they belong to.
```
Try to avoid ambiguous "backend", prefer explicit "renderer/rendering method" or "rendering driver", see https://github.com/godotengine/godot/pull/98744.

tetrapod00 commented on doc/classes/ProjectSettings.xml:
```diff
@@ -2861,6 +2861,9 @@
 		</member>
 		<member name="rendering/rendering_device/vulkan/max_descriptors_per_pool" type="int" setter="" getter="" default="64">
+			The number of descriptors per pool. Godot's Vulkan backend uses linear pools for descriptors that will be created and destroyed within a single frame. Instead of destroying every single descriptor every frame, they all can be destroyed at once by resetting the pool they belong to.
+			A larger number is more efficient up to a limit, after that it will only waste RAM (maximum efficiency is achieved when there is no more than 1 pool per frame). A small number could end up with one pool per descriptor, which negatively impacts performance.
```
```suggestion
			A larger number is more efficient up to a limit, after which it will only waste RAM (maximum efficiency is achieved when there is no more than 1 pool per frame). A small number could end up with one pool per descriptor, which negatively impacts performance.
```
Is there some guideline for actual numbers that are reasonable here, besides the default of 64? Does it depend on hardware?

darksylinc replied:
Hi, it's not HW dependent. The value of 64 is arbitrary.

Ideally this should not be necessary, see my [main post](https://github.com/godotengine/godot/pull/99257#issue-2660367614) when I refer to it:

> **Possible improvement for later:** It should be possible for Godot to adapt to how many descriptors per pool are needed on a per-key basis (i.e. grow their capacity like `std::vector` does) after rendering a few frames; which would be better than the current solution of having a single global value for all pools (`max_descriptors_per_pool`) that the user needs to tweak.

tetrapod00 replied:
Seems fine as-is then, if the default value is reasonable and the existing description describes the tradeoffs

tetrapod00 reviewed (commented):
Assuming correctness, the docs seem fine.

clayjohn commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp:
```diff
@@ -171,16 +171,18 @@ void RenderSceneBuffersRD::configure(const RenderSceneBuffersConfiguration *p_co
+
 	// Create our depth buffer.
-	create_texture(RB_SCOPE_BUFFERS, RB_TEX_DEPTH, get_depth_format(resolve_target, false, can_be_storage), get_depth_usage_bits(resolve_target, false, can_be_storage));
+	create_texture(RB_SCOPE_BUFFERS, RB_TEX_DEPTH, get_depth_format(resolve_target, false, can_be_storage), get_depth_usage_bits(resolve_target, false, can_be_storage) | extra_bits);
```
I don't think this can always be transient as we can end up reading from it in two situations:
1. The user reads from the depth texture in a GDShader
2. Various effects use it (SDFGI, VoxelGI, SSAO, SSIL, SSR, DoF etc.)

darksylinc replied:
Makes sense.

We should probably:

1. Add basic validation from our side to prevent invalid operations (even on non-tilers, so that users can catch the bug on their machines).
2. In another PR, evaluate whether the flag can be enabled. Leave this PR to add the functionality without setting the flag.

clayjohn replied:
That makes sense!

darksylinc replied:
Something about this was bothering me a lot, and finally put a huge improvement to it.

The problem is that one could ask for TRANSIENT flag erroneously, and not notice it is wrong until you deploy to Vulkan or Metal (and have Validation scream at you, or see that everything is broken).

This wrong usage needs to be detected on Desktop PCs, regardless of whether the GPU is TBDR.

I've added the following:

```cpp
#ifdef DEBUG_ENABLED
					if (((const TextureInfo *)uniform.ids[j * 2 + 1].id)->transient) {
						ERR_PRINT("TEXTURE_USAGE_TRANSIENT_BIT texture must not be used for sampling in a shader.");
					}
#endif
```

and:

```cpp
#ifdef DEBUG_ENABLED
	if (src_tex_info->transient) {
		ERR_PRINT("TEXTURE_USAGE_TRANSIENT_BIT p_src_texture must not be used in command_copy_texture.");
	}
	if (dst_tex_info->transient) {
		ERR_PRINT("TEXTURE_USAGE_TRANSIENT_BIT p_dst_texture must not be used in command_copy_texture.");
	}
#endif
```

Where needed.

The rationale is as follows:

1. It's behind DEBUG_ENABLED because in some cases (not all of them) it may be expensive enough to do this check in the exported game. But it should be done while in Editor builds (when it is paramount to detect it)
2. It must not fail / early out. The most likely scenario is that the user is running the Editor on a Desktop GPU (unless you're on an ARM mac). The GPU command will render fine. But the editor will show you the error. This is IMO the best user-friendly outcome: rendering happens the way user intended; but the user is notified something is wrong.
3. User will notice this error early in his big monster machine. Not late when he decides to deploy to a phone (and not all phones, e.g. RDNA phones are not tilers so the user won't notice anything wrong).

I've tested this and Mobile does not trigger the error, but Forward+ of course does; because it is as you say: SSAO and others need to sample from the Depth Buffer.

clayjohn replied:
Nice! I appreciate the extra error validation. 

We should still remove the flag before merging though

darksylinc commented:
I added TRANSIENT support for Metal too (it's called "Memoryless" there) but I did not test it on a real device.

clayjohn reviewed (commented):
Looks good to me. I have a few questions though:
1. Should we be trying to destroy shader modules as much as possible?
2. Should be using linear uniform sets as much as possible?
3. Should we use as many immutable samplers as possible?
4. For D3D12 and Metal, do we have a fallback for immutable samplers? I didn't notice any code that obviously included the immutable sampler

If the answer to any of 1, 2, or 3, is yes, then there are a lot of additions we can make in a follow up PR.

CC @dariosamo to take a look as well

darksylinc commented:
> Should we be trying to destroy shader modules as much as possible?

I'm unsure. Removing what's no longer necessary is often a good practice to save RAM. But we could ask TF if they found a problem in particular that motivated them to do it.

> Should be using linear uniform sets as much as possible?

"Yesn't".

Linear Uniform Sets (let's call it LUS and US for Uniform Set) are useful when we must recreate a US often.

Ideally we should try to create a US (or a collection of them) once and bind them when needed. Godot already tries to do this.

But there are times where, whether by the nature of the problem or a design issue, the US must be modified often which means re-creation. That's where LUS are useful.

There's still the case of "what happens when the LUS changes _sometimes_".

For example such is the case of the code being modified by this PR: `_update_render_base_uniform_set` was doing the following (simplified):

```cpp
void RenderForwardMobile::_update_render_base_uniform_set() {
    if( !render_base_uniform_set || is_dirty( render_base_uniform_set ) {
       recreate();
    }
}
```

That is, it only recreates if the uniforms haven't been created yet, or it is dirty and must be updated.

And now it has been changed to:

```cpp
void RenderForwardMobile::_update_render_base_uniform_set() {
    if( driver->use_linear_pools() || !render_base_uniform_set || is_dirty( render_base_uniform_set )  {
       recreate();
    }
}
```

That is, if the Godot driver (i.e. Vulkan backend) supports LUS, always recreate them.

The analysis here is deeper:

1. If the US is always dirty, it's a win.
2. If the US is never dirty, it's likely a loss. BUT, this is likely a simple scene so it won't matter.
3. If the US is somewhat dirty, then branch prediction goes into play. Unfortunately it would maybe be easier to if all driver backends supported LUS (if it even makes sense for the backend) so that there are no if statements: It always recreates. Sometimes the cost of the misprediction is equal or higher than the cost of just doing the operation.
    - Perhaps it's beneficial to always recreate (even in D3D12 & Metal) due to branch prediction woes.

**So the short version is:**

1. If the US requires recreating somewhat often, then prefer just recreating it every frame and use LUS.
2. If the US can be created once and almost never changes, then prefer not using LUS.

Note that this is why Vulkan has **_sets_** (at least 4 guaranteed by spec) that puts the "Set" in "Uniform Set": There are sets that will rarely change and will take 1 or 2 slots. And there are sets that will change every frame and should use the remaining slots.

Ideally one would use slot 0 for data that will never change and slot 1 for data that will always change (fun fact: D3D12 prefers the inverse: slot 0 for data that always changes). But reusability between PSOs creates reasons that some data (e.g. Material bindings) may have their own slot (despite that they could be put together with other static data).

> Should we use as many immutable samplers as possible?

On Desktop, it doesn't matter (unless we're using Border Color addressing mode on AMD, which is a rare addressing mode).
On Mobile (both iOS and Android) there are benefits.

Yes, we should try to Immutable samplers as much as possible. But also try to have as fewer samplers per PSO as possible.

This is hard to balance because defining more samplers (even if unused) allows for sharing the same Uniform Set with other PSOs as they become compatible.

Swapping US too much is costly on the CPU and trashes GPU caches. But having too many unused resources in the PSO increases the GPU cost.

> For D3D12 and Metal, do we have a fallback for immutable samplers? I didn't notice any code that obviously included the immutable sampler

We have a fallback and will use normal samplers. All 3 APIs support immutable samplers, but unfortunately don't agree how to define them:

1. Vulkan just wants us to have the `VkSampler` being passed alongside PSO creation.
2. D3D12 unfortunately doesn't want the equivalent of `VkSampler`, but rather the equivalent of `VkSamplerCreateInfo`. This needs a little more thinking to see if it's easy to adapt from what's been done, or requires further changes.
4. Metal wants the sampler defined in the shader code. Probably the best approach _if everyone did this too_. Since our native shader code is GLSL, this would require more analysis to see how to approach it.

clayjohn commented:
@darksylinc With respect to immutable samplers, I ask because we have these which are included in many shaders

https://github.com/godotengine/godot/blob/f952bfe9985ad8f507cc29b2c7601bbba18b8039/servers/rendering/renderer_rd/shaders/samplers_inc.glsl#L1-L12

I'm trying to figure out if it would make sense to unconditionally expose these through the PSO instead. 

From your description though, it sounds like it might be best to use immutable samplers for samplers that will _change_ between uniform sets so we can share uniform sets more

clayjohn commented:
Just to summarize the state of this PR so things don't get lost:
1. The transient flag needs to be removed from the depth resolve texture
2. This is not quite working on my M2 mac, so we need to debug why
3. The other comments are for follow up stuff, so they don't need any changes

darksylinc commented:
OK I've removed the flag, and as you found out; turns out that also fixed the M1 problem. Two birds one stone.

So I guess all that matters is approval & merge!

Btw the addition of TRANSIENT flag means this PR might cause merge conflicts with #98670

darksylinc commented:
Do NOT merge.

I spotted the shadow maps are inverted in certain situations. It's probably something wrong with the reverse Z patches I applied (TheForge made the switch to immutable samplers before Rev Z got merged).

darksylinc commented:
OK the bug was fixed.

As I suspected, I simply failed to apply the Reverse Depth correction to the immutable samplers. It's ready for review/merge again.

KeyboardDanni commented:
Am I correct in assuming that this would fix the additional frame of lag caused by waiting on the swapchain before submitting commands? Even with a frame queue size of 1, it looks like the existing behavior essentially ends up waiting on two V-Syncs before the results end up on-screen:

![2024-11-27 19_25_55-PIX on Windows - Timing 1 wpix](https://github.com/user-attachments/assets/e0267cf4-c633-4cc9-b7f2-71c2be6d9817)

By both creating and executing the commands before the upcoming V-blank, we'd end up saving a frame.

KeyboardDanni commented:
So I tested this PR. The new swapchain wait behavior *almost* works. It looks like it's still waiting on the next V-Blank before it submits the first command list (though the second one goes off without a hitch):

![2024-11-27 20_56_09-PIX on Windows - Timing 1 wpix](https://github.com/user-attachments/assets/b08556a9-504d-43f9-bd8f-cf6b879ad89c)
![2024-11-27 20_56_15-PIX on Windows - Timing 1 wpix](https://github.com/user-attachments/assets/cf7bda1f-c413-4896-ba05-7f7cbd1e8318)

darksylinc commented:
> Am I correct in assuming that this would fix the additional frame of lag caused by waiting on the swapchain before submitting commands?

No. This PR would eliminate pipeline bubbles between drawing and presenting, which results in lower performance.

Those measurements you're posting looks like something to send to NVIDIA. The capture you're posting shows the GPU literally doing nothing and waiting on nothing only to present two VBlanks later.

KeyboardDanni commented:
> No. This PR would eliminate pipeline bubbles between drawing and presenting, which results in improved performance thanks to the bubble removal.

Wouldn't this just move the bubble? The drawing and presenting were previously being done in one command list. Now it's being done in two, creating a bubble there, while removing the one between creating and submitting the command list. From your description:

> ## Do not wait so long for swapchain
> Godot currently does the following:
> 
> 1. Fill the entire cmd buffer with commands.
> 2. `submit()`
>    
>    * Wait with a semaphore for the swapchain.
>    * Trigger a semaphore to indicate when we're done (so the swapchain can submit).
> 3. `present()`
> 
> The optimization opportunity here is that 95% of Godot's rendering is done offscreen. Then a fullscreen pass copies everything to the swapchain. Godot doesn't practically render directly to the swapchain.
> 
> The problem with this is that the GPU has to wait for the swapchain to be released **to start anything**, when we could start _much earlier_. Only the final blit pass must wait for the swapchain.

Which mirrors what I'm seeing in these PIX traces. It's waiting until the swapchain is available, which doesn't happen until V-Sync. Only after the V-Sync marker you'll see the command list being executed. The proposed fix changes it to the following:

> 1. Fill the entire cmd buffer with commands.
> 2. In `screen_prepare_for_drawing` do `submit()`
>    
>    * There are no semaphore waits for the swapchain.
>    * Trigger a semaphore to indicate when we're done.
> 3. Fill a new cmd buffer that only does the final blit to the swapchain.
> 4. `submit()`
>    
>    * Wait with a semaphore for the submit() from step 2.
>    * Wait with a semaphore for the swapchain (so the swapchain can submit).
>    * Trigger a semaphore to indicate when we're done (so the swapchain can submit).
> 5. `present()`

In particular, the lack of wait on the swapchain in step 2.

> Those measurements you're posting looks like something to send to NVIDIA. The capture you're posting shows the GPU literally doing nothing and waiting on nothing only to present two VBlanks later.

But it's waiting on the swapchain, no? Aren't we controlling the synchronization primitives? Isn't that the purpose of APIs like Vulkan and D3D12?

KeyboardDanni commented:
@darksylinc Just to confirm, are we drawing to the same framebuffer every frame, and using that same framebuffer as the source when we copy it to the swapchain image? If so, it's likely that the framebuffer is locked waiting on the swapchain, meaning we still have to wait for an available swapchain image before we can run any commands. We might be able to fix this by having one framebuffer for each swapchain image, though if consistency of framebuffer contents between frames is important we'd have to add an extra copy, which might add some overhead.

darksylinc commented:
> Just to confirm, are we drawing to the same framebuffer every frame, and using that same framebuffer as the source when we copy it to the swapchain image? 

Yes.

>  If so, it's likely that the framebuffer is locked waiting on the swapchain, meaning we still have to wait for an available swapchain image before we can run any commands. 

No, because the swapchain was not acquired by the present engine during that time.
Rendering is roughly the following:

1. Draw into RTT0.
2. Wait until Swapchain Semaphore is released by the Present engine.
3. Copy RTT0 into Swapchain.
4. Present's engine must wait for the copy to release the execution semaphore so it can reacquire the Swapchain.
5. We render the next frame, to RTT0 (i.e. back to step 1 again). We only must wait on the semaphore released in step 3, not the one that was acquired in step 4.

The copy from step 3 can roughly take 0.5ms (give or take depending how fast the GPU is and screen resolution).
The presentation itself can take anywhere from 0 to 33.33ms to be released.

While present engine is still holding on to the Swapchain, we can start rendering to RTT0. Hopefully by the time we're done rendering to RTT0, the Swapchain was already released, causing no unnecessary bubbles.

If the swapchain hasn't yet been released, there will be a bubble. But it will be smaller than if we would have to wait just to even start rendering to RTT0.

KeyboardDanni commented:
> No, because the swapchain was not acquired by the present engine during that time.

But if all the images in the swapchain are currently waiting on the next V-Sync interval, how do we hand off the framebuffer? We can't reuse it until we perform the copy, and we need something to copy to.

> The copy from step 3 can roughly take 0.5ms (give or take depending how fast the GPU is and screen resolution). The presentation itself can take anywhere from 0 to 33.33ms to be released.

This is my main concern. In cases where the swapchain is full (which is going to be all the time if the GPU load is light), we cannot perform that copy until we acquire a new image. Because of this wait, we can't execute the command list for nearly a full V-Sync period. And because the command execution is delayed, we have to wait another V-Sync to get the final results on-screen instead of doing everything within the same V-Sync period.

> If the swapchain hasn't yet been released, there will be a bubble. But it will be smaller than if we would have to wait just to even start rendering to RTT0.

I'd think the bubble would be between executing the command list and presenting. The fact that it isn't, and command execution is still waiting on V-Sync (at least on my system) seems to imply that it's not quite working as expected.

KeyboardDanni commented:
Thinking about this some more, it's quite possible that this may just be solved with waitable swapchain, since we don't even bother to check if a swapchain image is available before we poll input and run game logic for the next frame. But we might get more throughput if we can ensure there's always an available framebuffer.

clayjohn commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4549,6 +4629,22 @@ void RenderingDevice::draw_list_draw_indirect(DrawListID p_list, bool p_use_indi
 	_check_transfer_worker_buffer(buffer);
 }
 
+void RenderingDevice::draw_list_set_viewport(DrawListID p_list, const Rect2 &p_rect) {
```
This is nice to have, but it seems like it is unused (I think it was originally part of the render pass optimizations right?). Let's leave it in at any rate as I want to use it for an optimization I have in mind for PointLight2Ds

darksylinc replied:
I looked into this, and you're right. This function of code was added by the Render Pass Optimization which was not included in this PR.

Anyway, soon another PR will be incoming w/ it.

clayjohn reviewed (approved):
Looks great! Let's get this merged ASAP so we get lots of testing in and don't risk another big conflict

arkology commented on servers/rendering/rendering_device.cpp:
```diff
@@ -3782,6 +3829,7 @@ Error RenderingDevice::screen_create(DisplayServer::WindowID p_screen) {
 Error RenderingDevice::screen_prepare_for_drawing(DisplayServer::WindowID p_screen) {
 	_THREAD_SAFE_METHOD_
 
+	// After submitting work, acquire the swapchain image(s)
```
```suggestion
	// After submitting work, acquire the swapchain image(s).
```

arkology commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4354,37 +4402,69 @@ void RenderingDevice::draw_list_draw(DrawListID p_list, bool p_use_indices, uint
 		}
 	}
 
+	// Bind the remaining batch
```
```suggestion
	// Bind the remaining batch.
```

arkology commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4873,37 +4969,70 @@ void RenderingDevice::compute_list_dispatch(ComputeListID p_list, uint32_t p_x_g
 		if (!cl->state.sets[i].bound) {
-			// All good, see if this requires re-binding.
-			draw_graph.add_compute_list_bind_uniform_set(cl->state.pipeline_shader_driver_id, cl->state.sets[i].uniform_set_driver_id, i);
+			// Descriptor set batching
```
```suggestion
			// Descriptor set batching.
```

arkology commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4873,37 +4969,70 @@ void RenderingDevice::compute_list_dispatch(ComputeListID p_list, uint32_t p_x_g
+			if (descriptor_set_batching) {
+				// All good, see if this requires re-binding.
+				if (i - last_set_index > 1) {
+					// If the descriptor sets are not contiguous, bind the previous ones and start a new batch
```
```suggestion
					// If the descriptor sets are not contiguous, bind the previous ones and start a new batch.
```

arkology commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4873,37 +4969,70 @@ void RenderingDevice::compute_list_dispatch(ComputeListID p_list, uint32_t p_x_g
+					valid_set_count = 1;
+					valid_descriptor_ids[0] = cl->state.sets[i].uniform_set_driver_id;
+				} else {
+					// Otherwise, keep storing in the current batch
```
```suggestion
					// Otherwise, keep storing in the current batch.
```

arkology commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4873,37 +4969,70 @@ void RenderingDevice::compute_list_dispatch(ComputeListID p_list, uint32_t p_x_g
 		}
 	}
 
+	// Bind the remaining batch
```
```suggestion
	// Bind the remaining batch.
```

arkology commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4986,37 +5115,68 @@ void RenderingDevice::compute_list_dispatch_indirect(ComputeListID p_list, RID p
 			// All good, see if this requires re-binding.
-			draw_graph.add_compute_list_bind_uniform_set(cl->state.pipeline_shader_driver_id, cl->state.sets[i].uniform_set_driver_id, i);
+			if (i - last_set_index > 1) {
+				// If the descriptor sets are not contiguous, bind the previous ones and start a new batch
```
```suggestion
				// If the descriptor sets are not contiguous, bind the previous ones and start a new batch.
```

arkology commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4986,37 +5115,68 @@ void RenderingDevice::compute_list_dispatch_indirect(ComputeListID p_list, RID p
+				valid_set_count = 1;
+				valid_descriptor_ids[0] = cl->state.sets[i].uniform_set_driver_id;
+			} else {
+				// Otherwise, keep storing in the current batch
```
```suggestion
				// Otherwise, keep storing in the current batch.
```

arkology commented on servers/rendering/rendering_device.cpp:
```diff
@@ -4986,37 +5115,68 @@ void RenderingDevice::compute_list_dispatch_indirect(ComputeListID p_list, RID p
 		}
 	}
 
+	// Bind the remaining batch
```
```suggestion
	// Bind the remaining batch.
```

arkology commented on servers/rendering/rendering_device.h:
```diff
@@ -191,7 +191,12 @@ class RenderingDevice : public RenderingDeviceCommons {
 
 	void update_perf_report();
-
+	// flag for batching descriptor sets
```
```suggestion
	// Flag for batching descriptor sets.
```

arkology reviewed (commented):
Stealing some of @AThousandShips job again  😄

AThousandShips commented:
Please don't tag me randomly though, it just creates a lot of unnecessary noise (also I had already made these corrections lol)

Repiteo reviewed (approved):
Expediting the merge process on this one; style nitpicks can be handled in a followup PR

Repiteo commented:
Thanks!
